internal/subscriber/retry/delayconsumer_test.go

```diff
@@ -81,6 +61,147 @@ func TestDelayConsumer_Run(t *testing.T) {
 	dc.Run(ctx)
 }
 
+func TestDelayConsumer_Pause_Resume(t *testing.T) {
+	ctx := context.Background()
+	subCtx, cancel := context.WithCancel(ctx)
+
+	ctrl := gomock.NewController(t)
+	mockBrokerStore := mocks.NewMockIBrokerStore(ctrl)
+	mockHandler := mocks3.NewMockMessageHandler(ctrl)
+	mockConsumer := mocks2.NewMockConsumer(ctrl)
+	cache := cachemock.NewMockICache(ctrl)
+
+	subs := getDummySubscriptionModel()
+	subscriberID := "subscriber-id"
+	mockBrokerStore.EXPECT().GetConsumer(gomock.Any(), gomock.Any()).Return(mockConsumer, nil)
+	mockConsumer.EXPECT().Resume(gomock.AssignableToTypeOf(subCtx), gomock.Any()).Return(nil).AnyTimes()
+	mockConsumer.EXPECT().Pause(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	// initialize delay consumer
+	errCh := make(chan error)
+	dc, err := NewDelayConsumer(subCtx, subscriberID, "t1", subs, mockBrokerStore, mockHandler, cache, errCh)
+	producer := buildMockProducer(ctx, ctrl)
+	assert.NotNil(t, dc)
+	assert.Nil(t, err)
+	assert.NotNil(t, dc.LogFields())
+	tFut := time.Now().Add(time.Second * 1)
+	msgs := make([]messagebroker.ReceivedMessage, 0)
+	msg := getDummyBrokerMessage()
+	msg.NextDeliveryTime = tFut
+	msgs = append(msgs, msg)
+	resp := &messagebroker.GetMessagesFromTopicResponse{Messages: msgs}
+	count := 0
+	mockConsumer.EXPECT().ReceiveMessages(gomock.Any(), gomock.Any()).DoAndReturn(
+		func(arg0 context.Context, arg1 messagebroker.GetMessagesFromTopicRequest) (*messagebroker.GetMessagesFromTopicResponse, error) {
+			if count == 0 {
+				count++
+				return resp, nil
+			} else {
+				return &messagebroker.GetMessagesFromTopicResponse{}, nil
+			}
+		}).AnyTimes()
+	cache.EXPECT().Get(gomock.Any(), gomock.Any()).Return([]byte{'0'}, nil).AnyTimes()
+	cache.EXPECT().Set(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	// on new message from broker
+	mockHandler.EXPECT().Do(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	mockConsumer.EXPECT().CommitByPartitionAndOffset(gomock.Any(), gomock.Any()).Return(messagebroker.CommitOnTopicResponse{}, nil).AnyTimes()
+	mockConsumer.EXPECT().Resume(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	mockBrokerStore.EXPECT().RemoveConsumer(gomock.Any(), gomock.Any()).Return(true)
+	mockBrokerStore.EXPECT().GetProducer(gomock.Any(), gomock.Any()).Return(producer, nil).AnyTimes()
+	mockConsumer.EXPECT().Close(gomock.Any()).Return(nil)
+	tests := []struct {
```

Comment:
we should also have a test case here wherein due to delay, message is not consumed

Reply:
Added message with 10s delay but total execution time 3s so there will be 1 remaining message

Reply:
we can have a delay in nanoseconds as well..